SwiftCompilerSources/Sources/Optimizer/Utilities/BorrowUtils.swift

```diff
@@ -0,0 +1,612 @@
+//===--- BorrowUtils.swift - Utilities for borrow scopes ------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2023 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
```

Comment:
It would be nice to add a little overview here which briefly explains the introduced concepts (mainly "borrowed value", "borrow introducer", "enclosing value", "inner/outer adjacent phi") with some small SIL examples. Let me try that (also to confirm if I understood everything correctly):

```
  %0 = some owned value      // borrow introducer/enclosing value of %1
  %1 = begin_borrow %0       // borrow introducer/enclosing value of %2
  %2 = begin_borrow %1       // borrow introducer/enclosing value of %3
  %3 = struct (%2)           // forwarding value
  
Borrow introducers (or enclosing values?) always dominate the value.
It is a little bit more complicated for reborrows (guaranteed phi arguments). The incoming values can be borrowed in not-dominating predecessor blocks. In this case that the borrow introducer is an adjacent phi argument in the same block:

  cond_br ..., bb1, bb2
bb1:
  %1 = some_owned_value
  %2 = begin_borrow %1
  br bb3(%1, %2)
bb2:
  %3 = some_owned_value
  %4 = begin_borrow %1
  br bb3(%3, %4)
bb3(%outer_adjacent_phi : @owned, %inner_adjacent_phi : @guaranteed):
  // %outer_adjacent_phi = borrow introducer of %inner_adjacent_phi
```

Reply:
Thanks. I already had comments for this purpose on a branch.

I just want to point out this pretty misleading
```
 %0 = some owned value      // borrow introducer/enclosing value of %1
  %1 = begin_borrow %0       // borrow introducer/enclosing value of %2
  %2 = begin_borrow %1       // borrow introducer/enclosing value of %3
  %3 = struct (%2)           // forwarding value
```
because for example `%1` is not the borrow introducer for `%2`.